### ISSUE TITLE
Connection attempts to unavailable brokers are not throttled

### ISSUE THREAD

**Issue by adamlamar** [Shopify/sarama#782]
##### Versions

*Please specify real version numbers or git SHAs, not just "Latest" since that changes fairly regularly.*
Sarama Version: Primarily running on 1.8.0, but the issue still appears to exist as of fd498173ae2bfc49a2b4eb19353241317785e7e6
Kafka Version: 0.8.2.2
Go Version: 1.5.3

##### Configuration

Using the async producer with a single topic (3 partitions, 2 replicas) and 3 brokers.

##### Logs

```
[11/08/16 04:59:58] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 04:59:58] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 04:59:58] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 04:59:58] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 04:59:58] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 04:59:58] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 04:59:58] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 04:59:58] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 04:59:58] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 04:59:58] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 04:59:58] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 04:59:58] [INFO] client.go:578: client/metadata retrying after 250ms... (3 attempts remaining)
[11/08/16 04:59:59] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 04:59:59] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 04:59:59] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 04:59:59] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 04:59:59] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 04:59:59] [INFO] client.go:578: client/metadata retrying after 250ms... (2 attempts remaining)
[11/08/16 04:59:59] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 04:59:59] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 04:59:59] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 04:59:59] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:00] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:00] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:00] [INFO] client.go:578: client/metadata retrying after 250ms... (1 attempts remaining)
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:00] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:00] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka4.example.com:9092
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:00] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:00] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka5.example.com:9092
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:00] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:00] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:00] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:00] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:00] [INFO] client.go:578: client/metadata retrying after 250ms... (3 attempts remaining)
[11/08/16 05:00:00] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:00] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka5.example.com:9092
[11/08/16 05:00:00] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:01] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:01] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:01] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:01] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:01] [INFO] client.go:578: client/metadata retrying after 250ms... (3 attempts remaining)
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka5.example.com:9092
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:01] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:01] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:01] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:01] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:01] [INFO] client.go:578: client/metadata retrying after 250ms... (2 attempts remaining)
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka5.example.com:9092
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:01] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:01] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:01] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:02] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:02] [INFO] client.go:578: client/metadata retrying after 250ms... (2 attempts remaining)
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:02] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:02] [INFO] client.go:578: client/metadata retrying after 250ms... (1 attempts remaining)
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:02] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:02] [INFO] client.go:578: client/metadata retrying after 250ms... (1 attempts remaining)
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:02] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:02] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:03] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:03] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
[11/08/16 05:00:03] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:410: client/brokers deregistered broker #-1 at kafka5.example.com:9092
[11/08/16 05:00:03] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka3.example.com:9092
[11/08/16 05:00:03] [INFO] broker.go:87: Failed to connect to broker kafka3.example.com:9092: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.182:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka4.example.com:9092
[11/08/16 05:00:03] [INFO] broker.go:87: Failed to connect to broker kafka4.example.com:9092: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.183:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:587: client/metadata fetching metadata for [events] from broker kafka5.example.com:9092
[11/08/16 05:00:03] [INFO] broker.go:87: Failed to connect to broker kafka5.example.com:9092: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:608: client/metadata got error from broker while fetching metadata: dial tcp 10.10.1.54:9092: getsockopt: connection refused
[11/08/16 05:00:03] [INFO] client.go:614: client/metadata no available broker to send metadata request to
[11/08/16 05:00:03] [INFO] client.go:419: client/brokers resurrecting 3 dead seed brokers
```

##### Problem Description

When _all_ kafka brokers become unavailable, sarama attempts to somewhat aggressively reconnect to each kafka broker. The rate appears to be driven by events like metadata refreshes as shown above. In our environment, each broker connection attempt spawns two DNS lookups (IPv4 and IPv6), and there is only a 250ms default wait between the three attempts. These attempts are made for each broker.

When using just a few sarama producers, this isn't a big issue, but as the number of producers rise, the connection attempts and associated DNS traffic can be excessive.

Ideally, sarama would do the following:

1. When a broker is unavailable, exponential backoff should be applied to ensure a reasonable number of connection attempts are made. If sarama has recently attempted a connection, it should wait until a penalty period has expired before reattempting.
2. Sarama should differentiate between connection failure and inability to refresh metadata (e.g. due to a leadership election in-progress), and not attempt reconnection if the broker is down.

I have a patch that addresses the first and possibly the second issue - would be happy to submit a PR against master if there is interest.

**Comment by eapache** [comment]
I am certainly happy to explore options for exponential backoff instead of linear.

There is also definitely room for optimization by coalescing calls to refresh metadata (currently each partition that fails will trigger its own refresh, which is simple but quite wasteful if many partitions fail simultaneously).

I am surprised that you're getting quite this level of flooding though, since the producer does have some circuit breakers built in. Repeat failures should open the breakers and trigger much more significant back-offs (10 seconds, if I recall correctly).

> Sarama should differentiate between connection failure and inability to refresh metadata (e.g. due to a leadership election in-progress), and not attempt reconnection if the broker is down.

It already does differentiate. When the failure is due to a connection error, it marks that broker as dead and retries on a different broker (normally it will just retry on the same broker after the backoff time). Only when *all* brokers are marked as dead does it start over with the seed set, and I'm not sure what else it can do in that scenario if it hasn't retried the requested number of times yet.

**Comment by adamlamar** [comment]
> I am certainly happy to explore options for exponential backoff

Sounds great! I'll try to submit a PR this week.

> quite this level of flooding though

The async producer definitely _does_ trigger the breaker and back off, but due to the retries and maybe other factors, it burns through a handful of connection attempts before backing off. The behavior isn't really that bad, we just have a large number of low volume producers and a fairly fragile infrastructure in this case that is sensitive to these sorts of bursts (which we don't have much control over, unfortunately).

> It already does differentiate

It sounds like we agree in the case where at least _one_ kafka broker is available. In the case you mentioned where _all_ brokers are unavailable, it seems clear from the above logs that `resurrecting 3 dead seed brokers` is happening fairly often, in the range of 1-3 times per second, and the metadata retries are causing connection attempts every 250ms (to the limit of the number of retries) despite no broker connection.

The patch I will submit still resurrects dead brokers, but brokers are aware of the last failed connection attempt and fail without a connection attempt if the exponential backoff timeout has not yet expired. This allows for per-broker backoff, and especially as the backoff period increases beyond the 10 second breaker timeout, one can easily see the connection attempts scatter.

**Comment by eapache** [comment]
> it seems clear from the above logs that `resurrecting 3 dead seed brokers` is happening fairly often, in the range of 1-3 times per second

Oh interesting, I took another look at this and the behaviour was not quite what I remembered it to be:
- if the update fails due to leadership election or other kafka-level issues, the client will wait the backoff amount and then retry on the same broker
- if the update fails due to network issues, the client will mark the broker as dead and *immediately* retry on another broker
- if all the brokers are dead, the client will wait the backoff amount and then resurrect all the brokers

So it's actually going to make `n` requests every 250ms (where `n` is the number of seed brokers). I had thought it was only making 1 request every 250ms.

Adding a back-off before retrying on a different broker might be sufficient to alleviate your main issue.

> still resurrects dead brokers, but brokers are aware of the last failed connection attempt and fail without a connection attempt if the exponential backoff timeout has not yet expired

Hmm, it might be simpler to just wait longer before resurrecting them in this case? Resurrecting them just so they can fail fast feels rather messy.
